packages/@react-aria/list/src/useListItem.ts

```diff
@@ -0,0 +1,76 @@
+import {getFocusableTreeWalker} from '@react-aria/focus';
+import {GridCollection} from '@react-types/grid';
+import {GridState} from '@react-stately/grid';
+import {HTMLAttributes, KeyboardEvent, RefObject} from 'react';
+import {KeyboardDelegate} from '@react-types/shared';
+import {MultipleSelectionManager} from '@react-stately/selection';
+import {Node} from '@react-types/shared';
+import {useGridCell} from '@react-aria/grid';
+
+interface ListItemOptions<> {
+  node: Node<unknown>,
+
+  selectionManager: MultipleSelectionManager,
+
+  keyboardDelegate: KeyboardDelegate,
+
+  ref: RefObject<HTMLElement>,
+
+  shouldFocusWrap?: boolean,
+
+  selectOnFocus?: boolean
+}
+
+// TODO
+interface ListItemAria {
+  listItemProps: HTMLAttributes<HTMLElement>
+}
+
+export function useListItem<T, C extends GridCollection<T>>(props: ListItemOptions, state: GridState<T, C>): ListItemAria {
+  let {
+    node,
+    ref
+  } = props;
+  let {gridCellProps} = useGridCell({
+    node,
+    ref,
+    focusMode: 'cell'
+  }, state);
+
+  let walker;
+  if (ref.current) {
+    walker = getFocusableTreeWalker(ref.current);
```

Comment:
I think you should create the tree walker inside `onKeyDown`, starting from `document.activeElement` rather than `ref.current`. This way it will be up to date in case focus moves between render and keyboard event. Generally not a good idea to access refs inside render as their value may change in the meantime.